File: ios/Sources/AdMobNextGenPlugin/rewardedinterstitial/RewardedInterstitialExecutor.swift
import Capacitor
import Foundation
import GoogleMobileAds

@objc
public class RewardedInterstitialExecutor: NSObject, FullScreenContentDelegate {

    private weak var plugin: AdMobNextGenPlugin?

    private var rewardedInterstitialAd: RewardedInterstitialAd?
    private var currentAdUnitId: String = ""

    private var lastLoadTime: TimeInterval = 0

    private var isRewardInterstitialEarned: Bool = false

    init(plugin: AdMobNextGenPlugin) {
        self.plugin = plugin
        super.init()
    }

    @objc public func loadAd(_ call: CAPPluginCall) {
        guard let plugin = self.plugin else { return }

        if !plugin.isInitialized {
            call.reject(
                "Google Mobile Ads SDK has not been initialized. Please call initialize() first."
            )
            return
        }

        let adUnitId = call.getString("adUnitId", "")
        if adUnitId.isEmpty {
            call.reject("Ad Unit ID is required.")
            return
        }

        let retryOpt = call.getDouble("retryInterval")
        let minLoadInterval: TimeInterval =
            (retryOpt != nil) ? (retryOpt! / 1000.0) : 5.0  
        let currentTime = Date().timeIntervalSince1970

        if (currentTime - lastLoadTime) < minLoadInterval {
            call.reject(
                "Request too fast. Please wait \(minLoadInterval * 1000)ms to prevent invalid traffic."
            )
            return
        }

        self.currentAdUnitId = adUnitId

        let request = Request()

        Task { @MainActor in
            do {
                self.rewardedInterstitialAd =
                    try await RewardedInterstitialAd.load(
                        with: adUnitId,
                        request: request
                    )
                self.rewardedInterstitialAd?.fullScreenContentDelegate = self

                self.rewardedInterstitialAd?.paidEventHandler = {
                    [weak self] value in
                    guard let self = self else { return }
                    var ret = JSObject()
                    ret["adUnitId"] = self.currentAdUnitId
                    ret["valueMicros"] = value.value.stringValue
                    ret["currencyCode"] = value.currencyCode
                    ret["precisionType"] = value.precision.rawValue
                    self.plugin?.notifyListeners(
                        "onRewardedInterstitialAdPaid",
                        data: ret
                    )
                }

                self.lastLoadTime = Date().timeIntervalSince1970  

                var ret = JSObject()
                ret["adUnitId"] = self.currentAdUnitId
                plugin.notifyListeners(
                    "onRewardedInterstitialAdLoaded",
                    data: ret
                )

                call.resolve()

            } catch {
                self.rewardedInterstitialAd = nil

                var ret = JSObject()
                ret["error"] = error.localizedDescription
                plugin.notifyListeners(
                    "onRewardedInterstitialAdFailedToLoad",
                    data: ret
                )

                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func showAd(_ call: CAPPluginCall) {
        guard let plugin = self.plugin else { return }

        if self.rewardedInterstitialAd == nil {
            call.reject("The rewarded interstitial ad is not ready yet.")
            return
        }

        DispatchQueue.main.async {
            guard let viewController = plugin.bridge?.viewController else {
                call.reject("Root ViewController not found.")
                return
            }

            if let ad = self.rewardedInterstitialAd {
                self.isRewardInterstitialEarned = false

                ad.present(from: viewController) { [weak self] in
                    self?.isRewardInterstitialEarned = true
                    let reward = ad.adReward
                    var ret = JSObject()
                    ret["amount"] = reward.amount.doubleValue
                    ret["type"] = reward.type
                    self?.plugin?.notifyListeners(
                        "onRewardedInterstitialAdReward",
                        data: ret
                    )
                }
                call.resolve()
            }
        }
    }

    public func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onRewardedInterstitialAdShowed", data: [:])
    }

    public func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        rewardedInterstitialAd = nil  
        if !self.isRewardInterstitialEarned {
            plugin?.notifyListeners("onRewardedInterstitialAdSkip", data: [:])
        }
        plugin?.notifyListeners("onRewardedInterstitialAdDismissed", data: [:])
    }

    public func ad(
        _ ad: FullScreenPresentingAd,
        didFailToPresentFullScreenContentWithError error: Error
    ) {
        rewardedInterstitialAd = nil  
        var ret = JSObject()
        ret["error"] = error.localizedDescription
        plugin?.notifyListeners(
            "onRewardedInterstitialAdFailedToShow",
            data: ret
        )
    }

    public func adDidRecordImpression(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onRewardedInterstitialAdImpression", data: [:])
    }

    public func adDidRecordClick(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onRewardedInterstitialAdClicked", data: [:])
    }
}

File: android/src/main/java/com/emi/plugins/admob/rewardedinterstitial/RewardedInterstitialExecutor.java
package com.emi.plugins.admob.rewardedinterstitial;

import android.app.Activity;
import androidx.annotation.NonNull;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.emi.plugins.admob.AdMobNextGenPlugin;
import com.emi.plugins.admob.ActionCallback;

import com.google.android.libraries.ads.mobile.sdk.common.AdLoadCallback;
import com.google.android.libraries.ads.mobile.sdk.common.AdRequest;
import com.google.android.libraries.ads.mobile.sdk.common.AdValue;
import com.google.android.libraries.ads.mobile.sdk.common.FullScreenContentError;
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError;
import com.google.android.libraries.ads.mobile.sdk.rewardedinterstitial.RewardedInterstitialAd;
import com.google.android.libraries.ads.mobile.sdk.rewardedinterstitial.RewardedInterstitialAdEventCallback;
import com.google.android.libraries.ads.mobile.sdk.rewarded.RewardItem;
import com.google.android.libraries.ads.mobile.sdk.rewarded.OnUserEarnedRewardListener;

public class RewardedInterstitialExecutor {

    private final AdMobNextGenPlugin plugin;
    private RewardedInterstitialAd mRewardedInterstitialAd;
    private String currentAdUnitId = "";
    private long lastLoadTime = 0; 

    public RewardedInterstitialExecutor(AdMobNextGenPlugin plugin) {
        this.plugin = plugin;
    }

    public void load(Activity activity, PluginCall call, final ActionCallback callback) {
        if (!plugin.isInitialized()) {
            callback.onError("Google Mobile Ads SDK has not been initialized. Please call initialize() first.");
            return;
        }
        String adUnitId = call.getString("adUnitId");
        if (adUnitId == null || adUnitId.isEmpty()) {
            callback.onError("Ad Unit ID is required.");
            return;
        }

        Double retryOpt = call.getDouble("retryInterval");
        long minLoadInterval = (retryOpt != null) ? retryOpt.longValue() : 5000L;
        long currentTime = System.currentTimeMillis();

        if ((currentTime - lastLoadTime) < minLoadInterval) {
            callback.onError("Request too fast. Please wait " + minLoadInterval + "ms to prevent invalid traffic.");
            return;
        }

        this.currentAdUnitId = adUnitId;

        activity.runOnUiThread(() -> {
            RewardedInterstitialAd.load(
                    new AdRequest.Builder(adUnitId).build(),
                    new AdLoadCallback<RewardedInterstitialAd>() {
                        @Override
                        public void onAdLoaded(@NonNull RewardedInterstitialAd ad) {
                            mRewardedInterstitialAd = ad;
                            lastLoadTime = System.currentTimeMillis(); 

                            JSObject ret = new JSObject();
                            ret.put("adUnitId", currentAdUnitId);
                            plugin.notifyPluginListeners("onRewardedInterstitialAdLoaded", ret);

                            callback.onSuccess();
                        }

                        @Override
                        public void onAdFailedToLoad(@NonNull LoadAdError adError) {
                            mRewardedInterstitialAd = null;

                            JSObject ret = new JSObject();
                            ret.put("error", adError.getMessage());
                            plugin.notifyPluginListeners("onRewardedInterstitialAdFailedToLoad", ret);
                            callback.onError(adError.getMessage());
                        }
                    }
            );
        });
    }

    public void show(Activity activity, ActionCallback callback) {
        if (mRewardedInterstitialAd == null) {
            callback.onError("The rewarded interstitial ad is not ready yet.");
            return;
        }

        activity.runOnUiThread(() -> {
            final boolean[] isRewardEarned = {false};
            mRewardedInterstitialAd.setAdEventCallback(new RewardedInterstitialAdEventCallback() {
                @Override
                public void onAdShowedFullScreenContent() {
                    plugin.notifyPluginListeners("onRewardedInterstitialAdShowed", new JSObject());
                }

                @Override
                public void onAdDismissedFullScreenContent() {
                    mRewardedInterstitialAd = null; 
                    if (!isRewardEarned[0]) {
                        plugin.notifyPluginListeners("onRewardedInterstitialAdSkip", new JSObject());
                    }
                    plugin.notifyPluginListeners("onRewardedInterstitialAdDismissed", new JSObject());
                }

                @Override
                public void onAdFailedToShowFullScreenContent(@NonNull FullScreenContentError error) {
                    mRewardedInterstitialAd = null; 
                    JSObject ret = new JSObject();
                    ret.put("error", error.getMessage());
                    plugin.notifyPluginListeners("onRewardedInterstitialAdFailedToShow", ret);
                }

                @Override
                public void onAdImpression() {
                    plugin.notifyPluginListeners("onRewardedInterstitialAdImpression", new JSObject());
                }

                @Override
                public void onAdClicked() {
                    plugin.notifyPluginListeners("onRewardedInterstitialAdClicked", new JSObject());
                }

                @Override
                public void onAdPaid(@NonNull AdValue value) {
                    JSObject ret = new JSObject();
                    ret.put("adUnitId", currentAdUnitId);
                    ret.put("valueMicros", value.getValueMicros());
                    ret.put("currencyCode", value.getCurrencyCode());
                    ret.put("precisionType", value.getPrecisionType().name());
                    plugin.notifyPluginListeners("onRewardedInterstitialAdPaid", ret);
                }
            });

            mRewardedInterstitialAd.show(activity, new OnUserEarnedRewardListener() {
                @Override
                public void onUserEarnedReward(@NonNull RewardItem rewardItem) {
                    isRewardEarned[0] = true;
                    JSObject ret = new JSObject();
                    ret.put("amount", rewardItem.getAmount());
                    ret.put("type", rewardItem.getType());
                    plugin.notifyPluginListeners("onRewardedInterstitialAdReward", ret);
                }
            });

            callback.onSuccess();
        });
    }
}
